docs/hugo/content/design/ADR-2023-05-Installing-Only-Selected-CRDs.md

```diff
@@ -0,0 +1,174 @@
+* Ensures that resources which are extremely likely to be needed (such as Microsoft.Resources/resourceGroup) are included 
+  by default.
+
+Cons:
+* Arbitrary.
+* May result in higher kube-apiserver memory consumption than is strictly needed for users as it is likely to include some
+  resources which they may not use.
+
+#### Answer 1b: We install nothing by default
+
+Pros:
+* Users choose exactly what they need.
+
+Cons:
+* The operator is nonfunctional out of the box, which is a poor new-user experience.
+
+#### Answer 1c: Selection of CRDs to install is required and the operator pod will not launch without it
+
+Can also use [Helm required values](https://stackoverflow.com/questions/53866196/how-best-to-say-a-value-is-required-in-a-helm-chart)
+to ensure that the value is required at the Helm chart level, although making it required may run afoul of requirement #3.
+
+Pros:
+* Users choose exactly what they need.
+
+Cons:
+* The operator is nonfunctional out of the box, but at least it's clearly nonfunctional and has a good error why.
+
+### Question 2: How do users specify what CRDs to install?
+
+#### Answer 2a: using our existing --crd-pattern command line parameter on the operator pod
+
+Pros:
+* Exists today
+
+Cons:
+* Command-line isn't preserved between upgrades, which fails requirement #3 (seamless upgrade experience).
+
+#### Answer 2b: Using a ConfigMap
+
+A `ConfigMap` would persist between upgrades and so in theory could be used to remember what was installed previously
+without the user needing to re-specify it.
+
+One challenge is that in the Helm context if the `ConfigMap` is created by Helm, Helm will overwrite it in 
+subsequent upgrades. This could be dealt with by using a 
+[pre-install hook](https://stackoverflow.com/questions/55503893/helm-patch-default-service-account). We would create and
+patch the `ConfigMap` in the pre-install hook.
+
+ConfigMap structure:
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: managed-resources
+  namespace: azureserviceoperator-system
+data:
+  "all": ""  # magic key meaning "*" which isn't supported in keys
+  microsoft.resources: "*"
+  microsoft.storage: "StorageAccount,StorageAccountsBlobService"
+```
+
+The following `patch` commands could be run either directly by the user or by our pre-install hook:
+
+* Addition of new resources: `kubectl patch configmap managed-resources -n azureserviceoperator-system --patch '{"data":{"microsoft.storage":"*"}}'`
+* Removal of old resources: `kubectl patch configmap managed-resources -n azureserviceoperator-system --type=json -p='[{"op": "remove", "path": "/data/microsoft.resources"}]'`
+
+If nothing is specified as part of the upgrade, the set of supported resources will not change.
+
+If the `ConfigMap` doesn't exist when we go to upgrade, we can assume we're coming from an older ASO version that didn't
+support the `ConfigMap` and default to creating one that signals to install all resources.
```

Comment:
This would put us back into the high memory use scenario.

Reply:
Yes but it'd be a breaking change IMO to do anything other than this, right?

Reply:
Yup. I just figure it's worth calling out as a Con for this approach.

Reply:
Note I'm specifically here referring to just upgrade scenarios. New installs user would need to choose still. I'll call that out as well.

Reply:
Added the con